### ISSUE TITLE
JEP on Windows with Python 3.5 requires a different setup.py file (included)

### ISSUE THREAD

**Issue by mrslezak** [mrj0/jep#80]
[setup.txt](https://github.com/mrj0/jep/files/1027616/setup.txt)

For normal Python 3.5 installs with Visual Studio 2015 (main) / 2013 (non-updated with Python 3.5) dependent components, this is the file to rename to setup.py before installing.  Otherwise you'll spend and hour figuring out all the include / library paths.  Builds on my PC.  I'd suggest Windows 7+ users use this setup.py instead of the supplied one to avoid installation issues.

**Comment by bsteffensmeier** [comment]
For Jep 3.7 I have been working on setting up automatic build/test of jep using Travis CI and AppVeyor. Using our default setup.py and the default windows python 3.5 installation I am able to build and test jep on AppVeyor. Are you saying this modified setup.py would always be necessary for python 3.5 or only under certain circumstances?

> 2013 (non-updated with Python 3.5) dependent components

I don't understand exactly what this means, do you have python components that were built with an older version of python?

**Comment by mrslezak** [comment]
No there are parts of Python 3.5 that require Windows 8.1 SDK files to
compile correctly.  Don't ask me why it just has those dependencies.  I
only use version 3.5.

**Comment by ndjensen** [comment]
When I've done individual tests with a virtual machine, it was always with Windows 7.  So I know the current version works with Windows 7.  I don't have access to my virtual machines at the moment, I don't remember which versions of Python 3 I was testing against.

Did you try following the [Windows build instructions here](https://github.com/mrj0/jep/wiki/Windows)?  Specifically for Python 3 it needs Microsoft Visual Studio 2010 aka MSVC 10.0, and you need the Windows 7 and .NET framework 4.  That ensures Jep is built with the same compiler as Python 3 itself (unless you built Python 3 from scratch or got it from somewhere other than [python.org](https://www.python.org/) in which case I don't know what happens).  I'm not sure what version of Windows AppVeyor is using, but it works there with Python 3.5.

Regarding your suggested change, I was specifically trying to avoid a dependency on setuptools since it doesn't come with Python.  Is setuptools ever going to be included with Python by default?  I know other projects don't care but right now we have zero dependencies on anything other than Python and Java.

Also you have hardcoded paths which isn't going to work for us.  Are you sure you need those paths after you got the Windows Kits installed?  We don't have to call those out at present if you have the development kit installed for Windows 7 and .NET framework 4.

**Comment by mrslezak** [comment]
It could be your computer isn't locked down as mine is on a company PC so
registry edits don't always go through.  The Windows build instructions
don't work for CPython 3.5 as it was built with Visual Studio 2015 and
plugins need the same compiler.  The hardcoded paths are default install
directories.  Maybe just have it available setup.py as an alternative for
those that need to install using: python setup.py install.

**Comment by bsteffensmeier** [comment]
[Staring in python 3.5](https://docs.python.org/3.5/whatsnew/3.5.html)

> Windows builds now use Microsoft Visual C++ 14.0, and extension modules should use the same.

So at the very least we need to update the windows instructions for 3.5 to reflect that. I do not have any windows machines to test on but I can verify that the current build setup works on AppVeyor, which [is using Windows Server 2012 R2](https://www.appveyor.com/docs/build-environment/).

If anyone has any other external documentation about why we would need changes it would be useful to link it here. I searched around a little and found some information [here](https://blog.ionelmc.ro/2014/12/21/compiling-python-extensions-on-windows/#for-python-3-5) that indicates you need to choose the right options when installing the Build Tools. Is it possible that @mrslezak installed the wrong SDK version and now needs the extra steps to work around that? If we provided the correct instructions on the wiki could we avoid changing setup.py?

**Comment by mrslezak** [comment]
I don't have external documents to back it up but on 5 PCs at work everyone
building CPython extensions needs files that are only available in the
earlier v8 SDK (even though Python docs say MSVC 14.0 only).  Windows 7 and
10.

**Comment by ndjensen** [comment]
I must have tested on Windows with Python 3.4.  I will see if I can replicate on Windows 7 with Python 3.5 before I attempt to go to Windows 8 or 10.  My personal laptop can't handle all these VMs with its smaller SSD hard drive.  Sooner or later we will need more thorough continuous integration or some platform testing volunteers, though TravisCI and Appveyor are a good step forward.  Thank you @mrslezak for reporting the issue as it does reveal issues with this platform.

I will attempt to get this fixed for Jep 3.7 and update the wiki so it can be seamless if you follow steps in the wiki.  At present I'm not sure how doable that is, but I somehow managed it on Windows 7 with the earlier versions of Python.  Like @bsteffensmeier suggested, if you or anyone finds or learns more, please share it with a comment on this ticket.

**Comment by ndjensen** [comment]
There is some information from Mike on the mailing list:

https://groups.google.com/d/msg/jep-project/xFZxg8n2X1c/rsZ5X8snDQAJ
https://groups.google.com/d/msg/jep-project/xFZxg8n2X1c/HFCZAQgaDAAJ

**Comment by ndjensen** [comment]
Ok, I just tried it on Windows 7 with Python 3.6 installed from python.org.  I had no issues.  I got the Visual C++ 2015 Build Tools (Standalone compiler, libraries and scriptscompiler) here: http://landinghub.visualstudio.com/visual-cpp-build-tools

When installing, I selected Windows 8.1 SDK and Windows 10 SDK.  After installing the build tools/SDKs, everything built and worked fine for me.

This page on the Python wiki is helpful: https://wiki.python.org/moin/WindowsCompilers.  Interestingly it implies the 2017 build tools require setuptools.  I have not tried the build with the 2017 tools.
